### PR TITLE
Add Admin API to refresh all workflow tasks

### PR DESCRIPTION
**Pull request by anish531213**

### CODE REVIEW

File: client/admin/metricClient.go [hunk 1]
```diff
@@ -256,3 +256,20 @@ func (c *metricClient) ReapplyEvents(
...
 	return err
 }
+
+func (c *metricClient) RefreshWorkflowTasks(
+	ctx context.Context,
+	request *shared.RefreshWorkflowTasksRequest,
+	opts ...yarpc.CallOption,
+) error {
+
+	c.metricsClient.IncCounter(metrics.FrontendRefreshWorkflowTasksScope, metrics.CadenceClientRequests)
```
Comment (wxing1292) [inline client/admin/metricClient.go]: AdminRefreshWorkflowTasksScope

File: client/history/metricClient.go [hunk 2]
```diff
@@ -594,3 +594,20 @@ func (c *metricClient) ReapplyEvents(
...
 	return err
 }
+
+func (c *metricClient) RefreshWorkflowTasks(
+	ctx context.Context,
+	request *h.RefreshWorkflowTasksRequest,
+	opts ...yarpc.CallOption,
+) error {
+
+	c.metricsClient.IncCounter(metrics.HistoryClientReapplyEventsScope, metrics.CadenceClientRequests)
```
Comment (wxing1292) [inline client/history/metricClient.go]: metrics name

File: service/frontend/adminHandler.go [hunk 3]
```diff
@@ -776,6 +776,35 @@ func (adh *AdminHandler) ReapplyEvents(
 	return nil
 }
 
+func (adh *AdminHandler) RefreshWorkflowTasks(
+	ctx context.Context,
+	request *gen.RefreshWorkflowTasksRequest,
+) (err error) {
+	defer log.CapturePanic(adh.GetLogger(), &err)
+	scope, sw := adh.startRequestProfile(metrics.AdminReapplyEventsScope)
```
Comment (wxing1292) [inline service/frontend/adminHandler.go]: metrics name

File: service/history/historyEngine.go [hunk 4]
```diff
@@ -2912,6 +2913,31 @@ func (e *historyEngineImpl) ReapplyEvents(
...
+	if err != nil {
+		return err
+	}
+	mutableStateTaskRefresher := newMutableStateTaskRefresher(
+		e.shard.GetConfig(),
+		e.shard.GetDomainCache(),
+		e.shard.GetEventsCache(),
+		e.shard.GetLogger(),
+	)
+	return mutableStateTaskRefresher.refreshTasks(e.shard.GetTimeSource().Now(), mutableState)
```
Comment (wxing1292) [inline service/history/historyEngine.go]: persistence to DB?

**Comment by yycptt** [comment]
Don't forget to also add an admin CLI command for this API.

File: service/history/historyEngine.go [hunk 5]
```diff
@@ -2912,6 +2913,43 @@ func (e *historyEngineImpl) ReapplyEvents(
...
+	)
+
+	now := e.shard.GetTimeSource().Now()
+
+	err = mutableStateTaskRefresher.refreshTasks(now, mutableState)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = mutableState.CloseTransactionAsMutation(now, transactionPolicyPassive)
```
Comment (yycptt) [inline service/history/historyEngine.go]: Why discard the results? `CloseTransactionAsMutation` will return a `WorkflowMutation` object which needs to be persisted. Also I think if passive policy is used, timer task for user timer and activity won't be generated.

File: service/history/historyEngine.go [hunk 6]
```diff
@@ -2912,6 +2913,43 @@ func (e *historyEngineImpl) ReapplyEvents(
...
+	mutableStateTaskRefresher := newMutableStateTaskRefresher(
+		e.shard.GetConfig(),
+		e.shard.GetDomainCache(),
+		e.shard.GetEventsCache(),
+		e.shard.GetLogger(),
+	)
+
+	now := e.shard.GetTimeSource().Now()
+
+	err = mutableStateTaskRefresher.refreshTasks(now, mutableState)
```
Comment (yycptt) [inline service/history/historyEngine.go]: If workflow has already closed, I don't think we need to refresh tasks.

File: service/history/historyEngine.go (hunk 5 above)
Comment (anish531213) [reply service/history/historyEngine.go]: fixed

File: client/admin/metricClient.go [hunk 7]
```diff
@@ -256,3 +256,20 @@ func (c *metricClient) ReapplyEvents(
...
 	return err
 }
+
+func (c *metricClient) RefreshWorkflowTasks(
+	ctx context.Context,
+	request *shared.RefreshWorkflowTasksRequest,
+	opts ...yarpc.CallOption,
+) error {
+
+	c.metricsClient.IncCounter(metrics.AdminRefreshWorkflowTasksScope, metrics.CadenceClientRequests)
```
Comment (yux0) [inline client/admin/metricClient.go]: AdminClientRefreshWorkflowTasksScope

File: service/history/historyEngine.go [hunk 8]
```diff
@@ -2912,6 +2913,55 @@ func (e *historyEngineImpl) ReapplyEvents(
...
+
+	mutableState, err := context.loadWorkflowExecution()
+	if err != nil {
+		return err
+	}
+
+	if !mutableState.IsWorkflowExecutionRunning() {
+		return nil
+	}
+
```
Comment (yux0) [inline service/history/historyEngine.go]: Also check if the domain is active, throw DomainNotActiveError if it is not.

File: service/history/historyEngine.go [hunk 9]
```diff
@@ -2912,6 +2913,55 @@ func (e *historyEngineImpl) ReapplyEvents(
...
+	)
+
+	now := e.shard.GetTimeSource().Now()
+
+	err = mutableStateTaskRefresher.refreshTasks(now, mutableState)
+	if err != nil {
+		return err
+	}
+
+	err = context.updateWorkflowExecutionWithNew(
```
Comment (yux0) [inline service/history/historyEngine.go]: Same as using updateWorkflowExecutionAsActive

File: go.mod [hunk 10]
```diff
@@ -40,6 +40,8 @@ require (
 	github.com/mattn/go-isatty v0.0.8 // indirect
 	github.com/mattn/go-runewidth v0.0.4 // indirect
 	github.com/mattn/go-sqlite3 v1.11.0 // indirect
+	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
+	github.com/modern-go/reflect2 v1.0.1 // indirect
```
Comment (yycptt) [inline go.mod]: Please run `go mod tidy` next time.

**Pull request merged**